### ISSUE TITLE
application specific chain: cosmos/tendermint, polkadot/substrate

### ISSUE THREAD

**Issue by Ivshti** [AdExNetwork/adex-core#12]
Porting the logic of `ADXExchange` into a cosmos app/zone makes a lot of sense, because: 

- instant finality
- more throughput
- dedicated zone only for the AdEx ecosystem
- ADX becomes a native staking token
- penalties for misbehaviour within the AdEx ecosystem, through slashing
- ability to do more complicated things, such as ring signatures, which can be used for restricting certain bids to groups of publishers, or signing bids as groups of advertisers
- running the nodes is already "taken care of", because the AdEx ecosystem requires running the `adex-node` anyway; being a validator could just be an option of the `adex-node`; Note that the `adex-node` is typically ran on an appropriate enviroment for a validator: highly available, powerful and secure hardware

The way this would happen is most likely with an airdrop to the Cosmos ecosystem, with a separately traded token. *Please note, exact token economics within Cosmos are still undecided*

This is much more simple and secure than having to go in/out between the ERC20 and the Cosmos Hub, and furthermore makes sense considering the *Cosmos version and Ethereum version will co-exist*. 

Given that the current dapp is implemented for Ethereum, and the Cosmos mainnet launch is months away, most real adoption will happen on Ethereum anyway.

Benefits of building on Cosmos rather than on Tendermint:

- consensus agnostic
- slashing/proof of stake 
- IBC, allowing it to connect to the Cosmos Hub
- authentication and coins (tokens) implemented in `x/auth` and `x/bank`

**Comment by Ivshti** [comment]
Things to yet figure out:

- how to ensure the implementation will be deterministic; most likely, we will implement it in go, so a static analyzer for such a thing would be great; a primitive (and naive) solution would be to just scan the code that handles the `sdk.Handler` and to warn when it uses possibly non-deterministic modules, external variables and other sources of non-determinism

**TO BE UPDATED**

**Comment by Ivshti** [comment]
Re token economy, we are chasing the following design goals/considerations

1) Zone fee token should be different than the zone staking token, in order to ensure the fee token is in constant circulation
2) AdEx will support any token for actual ad trading, so this part is irrelevant
3) relying on Peggy should be avoided, since it's not going to be ready any time soon (see ) 

So in conclusion, a plan that covers all those is

1) Keep ADX on Ethereum, allow it to be "moved" to cosmos once Peggy is ready
2) [Hard-spoon](https://blog.cosmos.network/introducing-the-hard-spoon-4a9288d3f0df) a new token on cosmos, from the ADX state, that will be used for staking and block provisions on the Zone
3) Allow any token for exchanging advertising or for paying transaction (bid action) fees

**Comment by Ivshti** [comment]
also see this:

https://medium.com/tendermint/proposed-cosmos-fee-token-codename-photon-e0927daf5c4c

https://blog.cosmos.network/introducing-the-hard-spoon-4a9288d3f0df

**Comment by Ivshti** [comment]
The way to hard-spoon ADX into a staking token on the Cosmos ecosystem, seems to be the following: create the AdEx zone with a map of all the ADX balances at a given snapshot block in the genesis state.

Then, the zone would mint the token (same balance as in the map) for the given user if they provide an Ethereum signed message containing their cosmos public key to the ClaimToken transaction.

Once minted, it can be transferred to the Cosmos Hub.

**Comment by Ivshti** [comment]
Another way of doing this, would be by deploying our own Ethermint zone rather than a spearately developed zone

Pros:

- Code re-use: using the same codebase the Ethereum implementation
- Easy upgradability
- Retaining the same plan of hard-spooned adex staking token, but possibly easier
- Allowing to do more complicated things, such as spliting bid rewards, etc. via smart contracts
- Developing in a deterministic language (solidity)

Cons:

- Gas fees might be harder to manage
- possible solidity pitfalls, especially when running on a different platform

**Comment by Ivshti** [comment]
Also worth adding here: https://blog.cosmos.network/why-application-specific-blockchains-make-sense-32f2073bfb37

**Comment by Ivshti** [comment]
Two more considerations:

- can our whole protocol be implemented over the built-in multisig transaction capability: it should be doable as long as those txes can timeout
- should a pegging system be part of the adex zone? i.e. the adex zone to also be a pegging zone; on first read this seems dangerous in terms of added complexity

**Comment by Ivshti** [comment]
nice example: https://github.com/livepeer/research/issues/8

**Comment by Ivshti** [comment]
Few more conclusions:

1. It seems that upgradability is something that will be handled better on Polkadot, because the entire chain runtime would be compiled to WASM and nodes will be able to seamlessly update to it
2. A specific usecase of validator slashing is the AdEx bid validators; esp in an open system where everybody can delegate their own validators, it makes sense to have public validators who've staked ADX

**Comment by Ivshti** [comment]
cool resources:

https://github.com/cosmos/cosmos-sdk/blob/develop/docs/sdk/core/app2.md
https://cosmos.network/docs/sdk/core/intro.html

**Comment by Ivshti** [comment]
more heplful stuff https://github.com/cosmos/cosmos-sdk/blob/develop/docs/sdk/sdk-by-examples/simple-governance/intro.md

**Comment by Ivshti** [comment]
regarding the security of tendermint/cosmos: https://github.com/cosmos/cosmos-sdk/commit/9a8f1a189a6e3d919b795c289ff083142866394c

**Comment by Ivshti** [comment]
an (almost) complete cosmos-sdk implementation is uploaded here: https://github.com/AdExNetwork/adex-protocol-cosmos

**Comment by thomasmodeneis** [comment]
Hi, 
May I suggest to deploy it into Smilo as well ?

**Comment by Ivshti** [comment]
ths issue is no longer relevant because the architecture is already firmly decided: primary contracts on Ethereum, with additional helper components built using substrate (potentially deployed on polkadot)
